aptos-move/framework/aptos-stdlib/sources/big_vector.move

```diff
@@ -0,0 +1,360 @@
+    }
+
+    /// Acquire an immutable reference to the `i`th element of the vector `v`.
+    /// Aborts if `i` is out of bounds.
+    public fun borrow<T>(v: &BigVector<T>, i: u64): &T {
+        assert!(i < length(v), error::invalid_argument(EINDEX_OUT_OF_BOUNDS));
+        vector::borrow(table_with_length::borrow(&v.buckets, i / v.bucket_size), i % v.bucket_size)
+    }
+
+    /// Return a mutable reference to the `i`th element in the vector `v`.
+    /// Aborts if `i` is out of bounds.
+    public fun borrow_mut<T>(v: &mut BigVector<T>, i: u64): &mut T {
+        assert!(i < length(v), error::invalid_argument(EINDEX_OUT_OF_BOUNDS));
+        vector::borrow_mut(table_with_length::borrow_mut(&mut v.buckets, i / v.bucket_size), i % v.bucket_size)
+    }
+
+    public fun append<T: store>(lhs: &mut BigVector<T>, other: BigVector<T>) {
+        let other_len = length(&other);
+        let half_other_len = other_len / 2;
+        let i = 0;
+        while (i < half_other_len) {
+            push_back(lhs, swap_remove(&mut other, i));
+            i = i + 1;
+        };
+        while (i < other_len) {
+            push_back(lhs, pop_back(&mut other));
+            i = i + 1;
+        };
+        destroy_empty(other);
+    }
+
+    /// Add element `val` to the end of the vector `v`. It grows the buckets when the current buckets are full.
+    /// This operation will cost more gas when it adds new bucket.
+    public fun push_back<T: store>(v: &mut BigVector<T>, val: T) {
+        if (v.end_index == v.num_buckets * v.bucket_size) {
+            table_with_length::add(&mut v.buckets, v.num_buckets, vector::empty());
+            v.num_buckets = v.num_buckets + 1;
+        };
+        vector::push_back(table_with_length::borrow_mut(&mut v.buckets, v.num_buckets - 1), val);
+        v.end_index = v.end_index + 1;
+    }
+
+    /// Pop an element from the end of vector `v`. It doesn't shrink the buckets even if they're empty.
+    /// Call `shrink_to_fit` explicity to deallocate empty buckets.
+    /// Aborts if `v` is empty.
+    public fun pop_back<T>(v: &mut BigVector<T>): T {
+        assert!(!is_empty(v), error::invalid_argument(EINDEX_OUT_OF_BOUNDS));
+        let last_bucket = table_with_length::borrow_mut(&mut v.buckets, v.num_buckets - 1);
+        let val = vector::pop_back(last_bucket);
+        // Shrink the table if the last vector is empty.
+        if (vector::is_empty(last_bucket)) {
+            move last_bucket;
+            vector::destroy_empty(table_with_length::remove(&mut v.buckets, v.num_buckets - 1));
+            v.num_buckets = v.num_buckets - 1;
+        };
+        v.end_index = v.end_index - 1;
+        val
+    }
+
+    /// Remove the element at index i in the vector v and return the owned value that was previously stored at i in v.
+    /// All elements occurring at indices greater than i will be shifted down by 1. Will abort if i is out of bounds.
+    public fun remove<T>(v: &mut BigVector<T>, i: u64): T {
```

Comment:
This can be impossibly heavy, I wonder if we should just avoid providing it.
(and same for all the O(n) algorisms below)

Reply:
I think for completeness, we may provide this but with comments stressing the cost. Otherwise ppl have to implement it by themselves in case they need it. I think it is different from enforcing any ordering of hash map since the semantic of those operations are deterministic, just whether to include an implementation.